bridge/sam/pool.go

```diff
@@ -0,0 +1,377 @@
+	p.tryToPromote(message.Hash)
+}
+
+// Add adds new message with the signature to pool
+func (p *pool) AddSignature(signature *MessageSignature) {
+	p.changeValidatorsLock.RLock()
+	defer p.changeValidatorsLock.RUnlock()
+
+	if p.hasConsumed(signature.Hash) {
+		// we no longer put the signature if the message has been consumed
+		return
+	}
+
+	p.messageSignatures.Put(signature)
+	p.tryToPromote(signature.Hash)
+}
+
+// ConsumeMessage sets the consumed flag and delete the message from pool
+func (p *pool) ConsumeMessage(hash types.Hash) {
+	p.consumedMap.Store(hash, true)
+
+	p.messageSignatures.RemoveMessage(hash)
+	p.messageMap.Delete(hash)
+	p.readyMap.Delete(hash)
+}
+
+// knows returns the flag indicating the message is known
+func (p *pool) isMessageKnown(hash types.Hash) bool {
+	_, ok := p.messageMap.Load(hash)
+	if !ok {
+		return false
+	}
+
+	return ok
+}
+
+// consumed returns the flag indicating the message is consumed
+func (p *pool) hasConsumed(hash types.Hash) bool {
+	raw, ok := p.consumedMap.Load(hash)
+	if !ok {
+		return false
+	}
+
+	consumed, ok := raw.(bool)
+
+	return ok && consumed
+}
+
+// GetReadyMessages returns the messages with enough signatures
+func (p *pool) GetReadyMessages() []ReadyMessage {
+	p.changeValidatorsLock.RLock()
+	defer p.changeValidatorsLock.RUnlock()
+
+	res := make([]ReadyMessage, 0)
+
+	p.readyMap.Range(func(key, value interface{}) bool {
+		hash, _ := key.(types.Hash)
+
+		if ready, _ := value.(bool); !ready {
+			return true
+		}
+
+		body := p.getMessageBody(hash)
+		if body == nil {
+			return true
+		}
+
+		signatures := p.messageSignatures.GetSignatures(hash)
+		res = append(res, ReadyMessage{
+			Body:       body,
+			Hash:       hash,
+			Signatures: signatures,
+		})
+
+		return true
+	})
+
+	return res
+}
+
+// UpdateValidatorSet update validators and threshold
+// This process blocks other processes because messages would lose the signatures
+func (p *pool) UpdateValidatorSet(validators []types.Address, threshold uint64) {
+	p.changeValidatorsLock.Lock()
+	defer p.changeValidatorsLock.Unlock()
+
+	oldValidators := p.validators
+	oldThreshold := p.signatureThreshold //nolint
+
+	p.validators = validators
+	atomic.StoreUint64(&p.signatureThreshold, threshold)
+
+	var maybeDemotableHashes []types.Hash
+	if removed := common.DiffAddresses(oldValidators, validators); len(removed) > 0 {
+		maybeDemotableHashes = p.messageSignatures.RemoveSignatures(removed)
+	}
+
+	if oldThreshold != threshold {
+		// we need to check all messages if threshold changes
+		p.tryToPromoteAndDemoteAll()
+	} else if len(maybeDemotableHashes) > 0 {
+		for _, hash := range maybeDemotableHashes {
+			p.tryToDemote(hash)
+		}
+	}
+}
+
+// canPromote return the flag indicating it's possible to change status to ready
+// message need to have enough signatures and be known by pool for promotion
+func (p *pool) canPromote(hash types.Hash) bool {
+	isKnown := p.isMessageKnown(hash)
```

Comment:
Why do we need to check both whether the message is known and if it has enough signatures?

Wouldn't it be enough just to check the signatures?

Reply:
To pick events up, you need to have data and enough signatures. Because expecting other validators to send hash and its signature by gossip. When validating, you can just check signature count.

Reply:
Thank you for the clarification, resolving this comment @Kourin1996